gunrock/app/bc/bc_enactor.cuh

```diff
@@ -441,45 +421,61 @@ struct BCBackwardIterationLoop : public IterationLoopBase
             (received_length, peer_, expand_op);
         return retval;
     }
-    
-    cudaError_t Gather(int peer_) {
 
-        auto &data_slice       = this -> enactor ->problem -> data_slices[this -> gpu_num][0];
-        auto &enactor_slice    = this -> enactor ->enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
+    cudaError_t Gather(int peer_) {
+        cudaError_t retval     = cudaSuccess;
+        auto &data_slice       = this -> enactor ->
+            problem -> data_slices[this -> gpu_num][0];
+        auto &enactor_slice    = this -> enactor ->
+            enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
         auto &enactor_stats    = enactor_slice.enactor_stats;
         auto &frontier         = enactor_slice.frontier;
         auto &oprtr_parameters = enactor_slice.oprtr_parameters;
 
         // printf("-- Gather --\n");
         // printf("  queue_length=%d\n", frontier.queue_length);
-        
+
         SizeT cur_pos = data_slice.forward_queue_offsets[peer_].back();
         data_slice.forward_queue_offsets[peer_].pop_back();
         SizeT pre_pos = data_slice.forward_queue_offsets[peer_].back();
         frontier.queue_reset  = true;
         if (enactor_stats.iteration > 0 && cur_pos - pre_pos > 0) {
             frontier.queue_length = cur_pos - pre_pos;
-            
-            // bool over_sized = false;
+
+            bool over_sized = false;
             // if (enactor_stats -> retval = Check_Size<SizeT, VertexId> (
-            //     enactor -> size_check, "queue1", 
-            //     frontier.queue_length, 
-            //     &frontier_queue -> keys[frontier_queue -> selector], 
+            //     enactor -> size_check, "queue1",
+            //     frontier.queue_length,
+            //     &frontier_queue -> keys[frontier_queue -> selector],
             //     over_sized, thread_num, enactor_stats->iteration, peer_, false)) return;
-            
-            MemsetCopyVectorKernel<<<120, 512, 0, oprtr_parameters.stream>>>(
-                frontier.V_Q()->GetPointer(util::DEVICE),
-                data_slice.forward_output[peer_].GetPointer(util::DEVICE) + pre_pos,
-                frontier.queue_length);
+            retval = CheckSize<SizeT, VertexT> (
+                (this -> enactor -> flag & Size_Check) != 0, "queue1",
+                frontier.queue_length, frontier.V_Q(),
+                over_sized, this -> gpu_num, enactor_stats.iteration, peer_);
+            if (retval)
+                return retval;
+
+            //MemsetCopyVectorKernel<<<120, 512, 0, oprtr_parameters.stream>>>(
+            //    frontier.V_Q()->GetPointer(util::DEVICE),
+            //    data_slice.forward_output[peer_].GetPointer(util::DEVICE) + pre_pos,
+            //    frontier.queue_length);
+            auto &forward_output = data_slice.forward_output[peer_];
+            GUARD_CU(frontier.V_Q()->ForAll([
```

Comment:
Are the various APIs for these `ForAll`/`ForEach` methods documented anywhere?

Reply:
They are part of the Array1D structure, https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/util/array_utils.cuh#L259 . Looks like there are too many variances, and I'm still not very sure which ones to keep. Based on current usages, I think just keeping the lambda and passing other arrays via lambda parameters, instead of via parameters to `ForAll` / `ForEach` functions, can be the way to go.